### PR TITLE
Fix race condition. Closes #2004

### LINKED ISSUE #2004: Race condition in updater
We `shell_execute` the updated client before we `CClient::Quit()` the old one, there is a chance that as the new one wants to read the config, the old one could be writing the config, nuking the old config as soon as you quit the new client.

This is because windows opens files exclusively. I think we should instead set a restart flag and only `shell_execute` right before the client shuts down.

### PR DESCRIPTION
**Pull request by Learath2**
I guess a new state is better then a global variable

### CODE REVIEW

File: src/engine/client.h [hunk 1]
```diff
@@ -71,6 +71,7 @@ class IClient : public IInterface
 		STATE_ONLINE,
 		STATE_DEMOPLAYBACK,
 		STATE_QUITING,
+		STATE_RESTARTING
```
Comment (def-) [inline src/engine/client.h]: , at the end

File: src/engine/client/client.cpp [hunk 2]
```diff
@@ -548,7 +548,7 @@ int *CClient::GetDirectInput(int Tick)
 // ------ state handling -----
 void CClient::SetState(int s)
 {
-	if(m_State == IClient::STATE_QUITING)
+	if(m_State >= IClient::STATE_QUITING)
```
Comment (def-) [inline src/engine/client/client.cpp]: Personally I'd be more comfortable with an explicit `m_State == STATE_QUITTING || m_State == STATE_RESTARTING`, then you don't need to remember the order, and it's not obvious to me that RESTARTING > QUITTING.

File: src/engine/client/client.cpp [hunk 3]
```diff
@@ -2360,7 +2358,7 @@ void CClient::PumpNetwork()
 	if(State() != IClient::STATE_DEMOPLAYBACK)
 	{
 		// check for errors
-		if(State() != IClient::STATE_OFFLINE && State() != IClient::STATE_QUITING && m_NetClient[CLIENT_MAIN].State() == NETSTATE_OFFLINE)
+		if(State() != IClient::STATE_OFFLINE && State() < IClient::STATE_QUITING && m_NetClient[CLIENT_MAIN].State() == NETSTATE_OFFLINE)
```
Comment (def-) [inline src/engine/client/client.cpp]: maybe also != STATE_QUITTING && != STATE_RESTARTING here

Comment (Learath2) [reply src/engine/client/client.cpp]: This one does kinda make sense though. If you really think it's an issue, I can change it.

**Pull request merged**